staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -911,20 +912,25 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 	}
 
 	if len(schema.XValidations) > 0 {
-		structural, err := structuralschema.NewStructural(schema)
-		if err != nil {
-			allErrs = append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), err))
+		if schemaIsNonStructural {
```

Comment:
We should just forbid `XValidations` for non-structural schemas, not even check them. This case can only happen when the CRD at hand survived from etcd through upgrades, i.e. when v1beta1 CRDs were still a thing. It would simplify this code a lot.

Reply:
+1

Reply:
`requireStructuralSchema` is used to ensure XValidations only used in structural schema. `requireStructuralSchema` will be false when oldCRD is non-structural(like v1beta1 CRDs) and we will return error if trying to update with XValidations existing without further checking.